File: SampleProjects/GoT-Characters-master/Objective-C/GoT/AsyncLoadConfiguration.m
//
//  LoaderConfiguration.m
//  GoT
//
//  Created by Paciej on 24/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "AsyncLoadConfiguration.h"

@interface AsyncLoadConfiguration ()

@property(nonatomic, copy, nonnull) id _Nullable (^parsingBlock)
(NSData *_Nonnull result);
@property(nonatomic, copy, nonnull) NSString *endpoint;
@property(nonatomic, copy, nullable) NSString *query;

@end

@implementation AsyncLoadConfiguration

- (nonnull instancetype)
initWithResponseParsingBlock:
(nonnull id _Nullable (^)(NSData *_Nonnull result))block
webserviceEndpoint:(nonnull NSString *)endpoint
webserviceQuery:(nullable NSString *)query {
    self = [super init];
    if (self) {
        self.parsingBlock = block;
        self.endpoint = endpoint;
        self.query = query;
    }
    return self;
}

- (nonnull id _Nullable (^)(NSData *_Nonnul))responseParsingBlock {
    return self.parsingBlock;
}

- (nonnull NSString *)webserviceEndpoint {
    return self.endpoint;
}

- (nullable NSString *)webserviceQuery {
    return self.query;
}

@end

@implementation AsyncLoadConfiguration (Article)
+ (nonnull instancetype)asyncLoadConfigurationFromArticle:
(nonnull Article *)article {
    return [[AsyncLoadConfiguration alloc]
            initWithResponseParsingBlock:^id _Nullable(NSData *_Nonnull result) {
                NSData *data;
                if ([result isKindOfClass:[NSData class]] &&
                    [UIImage imageWithData:result]) {
                    data = (NSData *)result;
                }
                return data;
            } webserviceEndpoint:article.thumbnailURLString
            webserviceQuery:nil];
}
@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/AsyncLoadConfiguration.swift
//
//  AsyncLoadConfiguration.swift
//  GoT
//
//  Created by Paciej on 24/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

//
//  LoaderConfiguration.m
//  GoT
//
//  Created by Paciej on 24/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import Foundation

class AsyncLoadConfiguration: NSObject {
    var responseParsingBlock: ((_ result: Data) -> Any?)? {
        return parsingBlock ?? { _ in return  }
    }
    var webserviceEndpoint: String {
        return endpoint
    }
    var webserviceQuery: String? {
        return query
    }

    private var parsingBlock: ((_ result: Data) -> Any?)!
    private var endpoint = ""
    private var query: String?

    init(responseParsingBlock block: @escaping (_ result: Data) -> Any?, webserviceEndpoint endpoint: String, webserviceQuery query: String?) {
        super.init()
        
        parsingBlock = block
        self.endpoint = endpoint
        self.query = query
    
    }
}

extension AsyncLoadConfiguration {
    convenience init(fromArticle article: Article) {
        self.init(responseParsingBlock: { result in
            var data: Data?
            if (result is Data) && UIImage(data: result) != nil {
                data = result
            }
            return data
        }, webserviceEndpoint: article.thumbnailURLString, webserviceQuery: nil)
    }
}